Add default auth model provider

hubert123490 opened the pull request:
My solution for default azure auth model provider. Please note that Azure advises to use different authentication method for production https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet.

> Simplifies authentication while developing apps that deploy to Azure by combining credentials used in Azure hosting environments with credentials used in local development. In production, it's better to use something else. See [Usage guidance for DefaultAzureCredential](https://aka.ms/azsdk/net/identity/credential-chains#usage-guidance-for-defaultazurecredential).

However it should be sufficient for development purposes and further collaboration. 

- Fixes: https://github.com/quarkiverse/quarkus-langchain4j/issues/1995

sberyozkin commented on model-providers/openai/openai-common/runtime/pom.xml:
```diff
@@ -26,6 +26,11 @@
             <artifactId>quarkus-langchain4j-core</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
```
Thanks @hubert123490 

I'm not sure this dependency belongs at the OpenAI common level... It should be at the Azure OpenAI level and the filter should be shipped there. Somehow it needs to replace the one checked at the OpenAI common level, may be by extending the QuarkusOpenAIClient, or being an Alternative CDI bean, it is a bit tricky but should be doable, thanks

hubert123490 commented:
@sberyozkin fair point
Im not quite happy abbout configMap, but it seems to work fine overall. There are some issues tho, like quite long token initialization time for the first request. Also it would be great for developer to be able to specify, which credential to use. Currently it gets first available from top to bottom (I think).
<img width="1501" height="184" alt="image" src="https://github.com/user-attachments/assets/36b46d58-4cfe-479f-b780-636882021e4d" />

sberyozkin commented on model-providers/openai/openai-common/runtime/src/main/java/io/quarkiverse/langchain4j/openai/common/DefaultOpenAiRestApiFilterResolver.java:
```diff
@@ -0,0 +1,16 @@
+import io.quarkiverse.langchain4j.auth.ModelAuthProvider;
+import io.quarkus.rest.client.reactive.QuarkusRestClientBuilder;
+
+public class DefaultOpenAiRestApiFilterResolver implements RestApiFilterResolver {
```
I'm just wondering, AFAIK, so far this filter can only be used with Azure OpenAI, I believe for direct OpenAI calls, OpenAI API Key is not passed as a bearer token, @geoand,  is it correct ?

Perhaps we can just move this filter out of OpenAI common ?

sberyozkin commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/AzureRestApiFilterResolver.java:
```diff
@@ -0,0 +1,27 @@
+import io.quarkiverse.langchain4j.openai.common.RestApiFilterResolver;
+import io.quarkus.rest.client.reactive.QuarkusRestClientBuilder;
+
+public class AzureRestApiFilterResolver implements RestApiFilterResolver {
```
I believe it has to become a synthetic CDI bean - you can see an example of how such beans are created in AzureOpenAI recorder - and you'd have the config properties available to the function that creates it

sberyozkin commented on model-providers/openai/openai-common/runtime/src/main/java/io/quarkiverse/langchain4j/openai/common/QuarkusOpenAiClient.java:
```diff
@@ -131,10 +132,13 @@ public MultivaluedMap<String, String> update(MultivaluedMap<String, String> inco
-                            .resolve(builder.configName)
-                            .ifPresent(modelAuthProvider -> restApiBuilder
-                                    .register(new OpenAiRestApi.OpenAIRestAPIFilter(modelAuthProvider)));
+                    Map<String, String> configMap = new HashMap<>();
```
I think the original code should stay

sberyozkin commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/AzureModelAuthProviderFilter.java:
```diff
@@ -0,0 +1,88 @@
+import io.quarkus.arc.InstanceHandle;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+
+public class AzureModelAuthProviderFilter implements ResteasyReactiveClientRequestFilter {
```
It should probably extend the Open AI common one

sberyozkin commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/AzureModelAuthProviderFilter.java:
```diff
@@ -0,0 +1,88 @@
+    private final ModelAuthProvider authorizer;
+
+    public AzureModelAuthProviderFilter() {
+        this.authorizer = new ApplicationDefaultAuthProvider();
```
We should make sure it is loaded only when no other custom model auth providers are available, for example, I linked earlier to the demo where an access token acquired after an Entra ID authorization code flow completion is propagated

sberyozkin commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/AzureModelAuthProviderFilter.java:
```diff
@@ -0,0 +1,88 @@
+            MultivaluedMap<String, Object> headers) implements ModelAuthProvider.Input {
+    }
+
+    private static class ApplicationDefaultAuthProvider implements ModelAuthProvider {
```
I wonder if we really need the filter resolver at all, and instead, only have this default model auth provider registered as a synthetic CDI bean in the if neither api or ad tokens or custom mode providers are available (this is the check that this PR removes but it should not)

hubert123490 commented:
I've tried to add synthetic bean for model auth provider
```
 @BuildStep
    @Record(ExecutionTime.RUNTIME_INIT)
    public void registerDefaultModelAuthProvider(AzureOpenAiRecorder recorder,
            BuildProducer<SyntheticBeanBuildItem> producer) {
        producer.produce(
                SyntheticBeanBuildItem
                        .configure(ModelAuthProvider.class)
                        .scope(ApplicationScoped.class)
                        .setRuntimeInit()
                        .createWith(recorder.modelAuthProvider())
                        .done());
    }
```
I added in azure recorder 
```
 public Function<SyntheticCreationalContext<ModelAuthProvider>, ModelAuthProvider> modelAuthProvider() {
        if (runtimeConfig.getValue().defaultConfig().apiKey().isPresent()
                || runtimeConfig.getValue().defaultConfig().adToken().isPresent()) {
            return new Function<>() {
                @Override
                public ModelAuthProvider apply(
                        SyntheticCreationalContext<ModelAuthProvider> modelAuthProviderSyntheticCreationalContext) {
                    return null;
                }
            };
        } else {
            return new Function<>() {
                @Override
                public ModelAuthProvider apply(
                        SyntheticCreationalContext<ModelAuthProvider> modelAuthProviderSyntheticCreationalContext) {
                    return new AzureOpenAiRecorder.ApplicationDefaultAuthProvider();
                }
            };
        }
    }
```

but that resulted in 
Error
"Null contextual instance was produced by a normal scoped synthetic bean: SYNTHETIC bean [class=io.quarkiverse.langchain4j.auth.ModelAuthProvider, id=FIym7E17BGsirk4ooWp8Cr-hKM0]"
when api key or ad token is provided
and no matter what I did, I couldn't create this bean conditionally.

However I was successful with BuildConfig approach, such as I added 

> quarkus.langchain4j.azure-openai.azure-default-credentials-enabled=true

```
@ConfigRoot(phase = BUILD_TIME)
@ConfigMapping(prefix = "quarkus.langchain4j.azure-openai")
public interface LangChain4jAzureOpenAiBuildConfig {

    /**
     * Chat model related settings
     */
    ChatModelBuildConfig chatModel();

    /**
     * Embedding model related settings
     */
    EmbeddingModelBuildConfig embeddingModel();

    /**
     * Moderation model related settings
     */
    ModerationModelBuildConfig moderationModel();

    /**
     * Image model related settings
     */
    ImageModelBuildConfig imageModel();

    /**
     * If true, enables the use of Azure Default Credentials for authentication.
     * If not set, the default behavior is disabled.
     */
    @ConfigDocDefault("true")
    Optional<Boolean> azureDefaultCredentialsEnabled();
}
```

and in processor
```
    @BuildStep
    @Record(ExecutionTime.RUNTIME_INIT)
    public void registerDefaultModelAuthProvider(AzureOpenAiRecorder recorder,
            BuildProducer<SyntheticBeanBuildItem> producer,
            LangChain4jAzureOpenAiBuildConfig buildConfig) {

        if (buildConfig.azureDefaultCredentialsEnabled().isPresent() &&
                buildConfig.azureDefaultCredentialsEnabled().get()) {
            producer.produce(
                    SyntheticBeanBuildItem
                            .configure(ModelAuthProvider.class)
                            .scope(ApplicationScoped.class)
                            .setRuntimeInit()
                            .createWith(recorder.modelAuthProvider())
                            .done());
        }
    }
```

And that approach I've already pushed. Since it's default bean it also can be overwritten. Only downside is that even when api-key is provided with quarkus.langchain4j.azure-openai.azure-default-credentials-enabled, the azure model auth provider is used, but maybe it can be handled by warning or exception. Anyways please let me know if it is acceptable one.

geoand commented:
It seems like the native failure is related

sberyozkin commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/runtime/AzureOpenAiRecorder.java:
```diff
@@ -281,6 +288,16 @@ public ImageModel apply(SyntheticCreationalContext<ImageModel> context) {
+            @Override
+            public ModelAuthProvider apply(
+                    SyntheticCreationalContext<ModelAuthProvider> modelAuthProviderSyntheticCreationalContext) {
+                return new AzureOpenAiRecorder.ApplicationDefaultAuthProvider();
```
It is really all that should be done 1) get the default provider registered as a bean 2) remove the check from the other function and move to your new function, but instead of throwing an exception, register the default one only if no other auth option exists

sberyozkin commented:
Thanks @hubert123490, let's take it further from this point, I'll try to  experiment as well with your PR, we just need to move that `throw` check that checks the tokens and a custom provider to your new function and register this provider only if if that check returns false

hubert123490 commented:
After several attempts, I believe registering a synthetic bean for DefaultModelAuthProvider correctly is not possible due to the CDI nature of Quarkus — or at least I was not able to find a way to do it. Here are my observations:

1. The synthetic bean must always be registered. This means that when an AD token, API key, or custom ModelAuthProvider is present, we are forced to return a null instance of DefaultModelAuthProvider. The problem is that later in the OpenAI common module, this null instance is picked up and registered as the ModelAuthProvider.

2. ModelAuthProvider has higher priority than API key and AD token in the resolution chain. So when the default bean is registered — even with a null instance — it takes priority and the API key and AD token are never reached.

The only real fixes would be to either pass the ModelAuthProvider directly into the OpenAI common module, or change the REST API filter logic — both of which are probably best avoided.

Given all of this, I suggest we register this bean based on build-time config instead. This way we keep things simple and still give the user a choice to opt into using the default Azure ModelAuthProvider.

This also makes sense from a practical standpoint. A developer running locally would typically want to use DefaultAzureCredentialBuilder, which conveniently picks up credentials from the environment, CLI, or IDE. However, in production they may want to restrict this to a specific credential type — for example, workload identity only — for security or compliance reasons. A build-time flag gives them exactly that control without any extra complexity.

I would really appreciate it if you could take a moment to consider this approach.

geoand commented on model-providers/openai/azure-openai/deployment/src/main/java/io/quarkiverse/langchain4j/azure/openai/deployment/AzureOpenAiProcessor.java:
```diff
@@ -8,13 +8,12 @@
-import org.jboss.jandex.AnnotationInstance;
-import org.jboss.jandex.ClassType;
-import org.jboss.jandex.ParameterizedType;
-import org.jboss.jandex.Type;
+import org.jboss.jandex.*;
```
Please avoid star imports

hubert123490 replied:
ok

geoand commented on model-providers/openai/azure-openai/deployment/src/main/java/io/quarkiverse/langchain4j/azure/openai/deployment/LangChain4jAzureOpenAiBuildConfig.java:
```diff
@@ -28,4 +29,10 @@ public interface LangChain4jAzureOpenAiBuildConfig {
+     * If true, enables the use of Azure Default Credentials for authentication.
+     */
+    @WithDefault("false")
+    boolean azureDefaultCredentialsEnabled();
```
Is this the proper name?

hubert123490 replied:
Changed to more appropriate one in next commit (atleast I think so :D)

geoand commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/runtime/AzureOpenAiRecorder.java:
```diff
@@ -367,4 +382,24 @@ public void run() {
     }
 
+    public static class ApplicationDefaultAuthProvider implements ModelAuthProvider {
+        private static final String SCOPE = "https://cognitiveservices.azure.com/.default";
+        private static final DefaultAzureCredential credential = new DefaultAzureCredentialBuilder().build();
+        private static final TokenRequestContext context = new TokenRequestContext().addScopes(SCOPE);
+        private static volatile AccessToken currentToken;
```
This will probably cause problems in native mode.

hubert123490 replied:
Using quarkus extension for azure identity should resolve any native build issues. https://github.com/quarkiverse/quarkus-azure-services/tree/main/common/azure-identity

geoand commented on model-providers/openai/azure-openai/runtime/src/main/java/io/quarkiverse/langchain4j/azure/openai/runtime/AzureOpenAiRecorder.java:
```diff
@@ -393,4 +407,24 @@ public void run() {
     }
 
+    public static class DefaultAzureCredentialModelAuthProvider implements ModelAuthProvider {
+        private static final String SCOPE = "https://cognitiveservices.azure.com/.default";
+        private static final TokenCredential credential = new DefaultAzureCredentialBuilder().build();
+        private static final TokenRequestContext context = new TokenRequestContext().addScopes(SCOPE);
```
This is interesting and makes me wonder whether any of this stuff needs to be configurable. If so, it also affects how we support configuration

geoand commented:
Can you please squash the commits and rebase onto the latest `main`?

geoand commented on model-providers/openai/azure-openai/deployment/src/main/java/io/quarkiverse/langchain4j/azure/openai/deployment/LangChain4jAzureOpenAiBuildConfig.java:
```diff
@@ -28,4 +29,10 @@ public interface LangChain4jAzureOpenAiBuildConfig {
+     * If true, provides model auth provider, which uses Default Azure Credential for authentication.
+     */
+    @WithDefault("false")
+    boolean defaultAzureCredentialModelAuthProvider();
```
```suggestion
    boolean useAzureCredentialModelAuthProvider();
```

hubert123490 committed:
Add azure default credential to azure openai

geoand reviewed (commented):
LGTM, thanks!

@sberyozkin WDYT?
